### DISCUSSION TITLE
Info: UX report for dtg/dtg-lib crates

### DISCUSSION THREAD

**Discussion by qtfkwk** [BurntSushi/jiff#43] [Experience Reports]
If interested, I just replaced chrono and chrono-tz dependencies in my dtg and dtg-lib crates without issue!

* https://github.com/qtfkwk/dtg
* [Commit](https://github.com/qtfkwk/dtg/commit/1068f137e3c6fb1b6e3e6d1cbe2652768635ef21)
* [Changes to `lib/src/lib.rs`](https://github.com/qtfkwk/dtg/commit/1068f137e3c6fb1b6e3e6d1cbe2652768635ef21#diff-f8cd0fc9fbbb8070d21e750fd63b08f122bc699b3589aedb677c63c1b65c2234)
* [Changes to `cli/src/main.rs`](https://github.com/qtfkwk/dtg/commit/1068f137e3c6fb1b6e3e6d1cbe2652768635ef21#diff-37970e7d817d4316a093bb967d74ff7e26d9c6c01d93f5a79d3d195ba9f6155b)

Notable observations include a maximum timestamp of 253402207200 (9999-12-30T22:00:00Z); using chrono it was 8210266876799 (262142-12-31T23:59:59Z); but no deal breakers.

Please feel free to close this issue.

**Comment by BurntSushi** [comment]
Nice! I think you might be the first. :-)

Did you hit any problems during the migration? Did you find anything easier? Harder?

**Reply by qtfkwk** [reply]
tldr: the migration was pretty straightforward... would say it is slightly easier with jiff

Summary of changes I needed to apply from chrono to jiff:

* Replace `use chrono::{offset::LocalResult, DateTime, Datelike, TimeZone, Timelike, Utc};` with `pub use jiff::{tz::TimeZone, Timestamp};` (pub is due largely to needing these exposed for doctests/tests w/o another `use`)
* Replace `DateTime<Utc>` with `Timestamp`
* Replace `Utc::now()` with `Timestamp::now()`
* Replace `Utc.timestamp_opt` with `Timestamp::new`
* Replace `if let LocalResult::Single(dt) = Utc.with_ymd_and_hms(y, v[4], v[3], v[2], v[1], v[0])` with `if let Ok(dt) = format!("{y}-{:02}-{:02}T{:02}:{:02}:{:02}Z", v[4], v[3], v[2], v[1], v[0]).parse::<Timestamp>()`
* Replace `Tz` with `TimeZone`
* Replace `Tz::UTC` with `TimeZone::UTC`
* Replace `.format` with `.strftime`
* Replace `.with_timezone(tz)` with `.to_zoned(tz.clone())`
* Replace `.month0()` with `.month() as u8 - 1`
* Replace `.day0()` with `.day() as u8 -1`
* Replace `s.parse()` with `jiff::tz::db().get(s)`
* Replace `TZ_VARIANTS.iter()` with `jiff::tz::db().available()`

Thanks, btw for this! (I'd been waiting/hoping for a newer datetime lib since chrono doesn't include timezones without chrono-tz, get the local timezone without iana-time-zone, and had [security issues](https://github.com/chronotope/chrono/issues?q=is%3Aissue+rustsec+is%3Aclosed).)

**Reply by BurntSushi** [reply]
> `if let Ok(dt) = format!("{y}-{:02}-{:02}T{:02}:{:02}:{:02}Z", v[4], v[3], v[2], v[1], v[0]).parse::<Timestamp>()`

For this, I think you can do:

```rust
let result = jiff::civil::Date::new(y, v[4], v[3]).and_then(|d| {
    Ok(d.to_datetime(jiff::civil::Time::new(v[2], v[1], v[0], 0)?))
}).and_then(|dt| dt.to_zoned(TimeZone::UTC)).map(|zdt| zdt.timestamp());
if let Ok(ts) = result { ... }
```

Although that is a bit of a mouthful! I'd probably split it out into a helper function so it could be written in a bit more of a linear fashion. But I do wonder if perhaps there should be some helper constructors on `Zoned` for cases like this.

**Reply by qtfkwk** [reply]
Cool! Applied it in dtg-lib [6.1.0](https://github.com/qtfkwk/dtg/commit/5c10919b034e487bd074b342073163abeaa64d99#diff-f8cd0fc9fbbb8070d21e750fd63b08f122bc699b3589aedb677c63c1b65c2234R158) as a new `Dtg::from_ymd_hms` function... it isn't too bad after cleaning up my variables, etc. Thanks!

**Comment by lukehsiao** [comment]
FWIW: I also made the switch in `openring-rs`: https://github.com/lukehsiao/openring-rs/commit/485fe4ef480c9f08a1c895b9e6b75b8c2b6f3774

I wasn't using chrono extensively, and found it very straightforward.

**Reply by BurntSushi** [reply]
w00t! Thanks for the feedback. :-)
